### PR TITLE
Universal listings - Add 'has child pages' filter

### PR DESCRIPTION
**Pull request by gasman**

### CODE REVIEW

File: wagtail/admin/views/pages/listing.py [hunk 1]
```diff
@@ -71,10 +86,26 @@ class PageFilterSet(WagtailFilterSet):
...
 
     class Meta:
         model = Page
-        fields = ["content_type", "latest_revision_created_at", "owner", "site"]
+        fields = [
+            "content_type",
+            "latest_revision_created_at",
+            "owner",
+            "site",
+            "has_child_pages",
```
Comment (laymonage) [inline wagtail/admin/views/pages/listing.py:107]: We probably shouldn't add this here because it's not an actual field.

django-filter will normally complain with `TypeError: 'Meta.fields' must not contain non-model field names: has_child_pages`, but in this case it's obscured because the filter is overridden by the `has_child_pages = HasChildPagesFilter(...)` above.

While it's unlikely, django-filter might be more aggressive in the future and enforce this even if we have defined a filter override. This is why I suggested using `fields = []` unless we want to auto-generate some filters from the fields :D

File: wagtail/admin/views/pages/listing.py [hunk 2]
```diff
@@ -45,6 +49,17 @@ def get_filter_predicate(self, v):
         return {"path__startswith": v.root_page.path}
 
 
+class HasChildPagesFilter(BooleanFilter):
```
Comment (laymonage) [inline wagtail/admin/views/pages/listing.py:52]: This works well, although I wonder if we should also special-case `BooleanFilter` in `BaseListingView.active_filters` so instead of showing `True`/`False`, we show "Yes"/"No" like the form does.

e.g. by adding

```py
            elif isinstance(filter_def, BooleanFilter):
                filters.append(
                    ActiveFilter(
                        filter_def.label,
                        {True: _("Yes"), False: _("No"), None: _("Any")}[value],
                        self.get_url_without_filter_param(field_name),
                    )
                )
```

Or alternatively, using a `ChoiceFilter`/`TypedChoiceFilter` instead, so we don't have to write the "Yes"/"No" code twice.

File: wagtail/admin/views/pages/listing.py [hunk 3]
```diff
@@ -71,10 +86,26 @@ class PageFilterSet(WagtailFilterSet):
         queryset=Site.objects.all(),
         widget=CheckboxSelectMultiple,
     )
+    has_child_pages = HasChildPagesFilter(
+        label=_("Has child pages"),
+        widget=Select(
```
Comment (laymonage) [inline wagtail/admin/views/pages/listing.py:91]: Any reason why not use a `RadioSelect` instead? Seems to be more intuitive that way IMO, one less click to apply the filter.

**Review by laymonage — COMMENTED** [review]
Works well, I just have a few minor comments. Thanks!

**Comment by gasman** [comment]
Thanks @laymonage! I like the idea of using ChoiceFilter - that saves having to fight against the way BooleanFilter wanted to do things :-) All updated now.

File: wagtail/admin/views/generic/base.py [hunk 4]
```diff
@@ -251,7 +251,11 @@ def active_filters(self):
 
         for field_name in self.filters.form.changed_data:
             filter_def = self.filters.filters[field_name]
-            value = self.filters.form.cleaned_data[field_name]
+            try:
+                value = self.filters.form.cleaned_data[field_name]
+            except KeyError:
+                continue  # invalid filter value
```
Comment (laymonage) [inline wagtail/admin/views/generic/base.py:257]: Thanks! I was working on this for a separate PR, but I'm happy to get this in here. I'll add the test as I merge.

File: wagtail/admin/tests/pages/test_explorer_view.py [hunk 5]
```diff
@@ -665,6 +665,58 @@ def test_filter_by_site(self):
...
+        self.assertEqual(
+            page_ids, {self.child_page.id, self.old_page.id, new_page_child.id}
+        )
+        self.assertContainsActiveFilter(
+            response,
+            "Has child pages: No",
+            "has_child_pages=false",
+        )
+
+    def test_invalid_filter(self):
```
Comment (laymonage) [inline wagtail/admin/tests/pages/test_explorer_view.py:704]: Test for the invalid filter fix, and I split the fix into a separate commit. Hope that's OK!

**Review by laymonage — APPROVED** [review]
